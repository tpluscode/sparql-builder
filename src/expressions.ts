export { IN } from './lib/expressions/in'
export { VALUES } from './lib/expressions/values'
