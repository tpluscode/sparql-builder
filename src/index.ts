export { sparql } from '@tpluscode/rdf-string'
export { DELETE } from './lib/DeleteBuilder.js'
export type { DeleteInsertQuery as DeleteInsert, DeleteData } from './lib/DeleteBuilder.js'
export { INSERT } from './lib/InsertBuilder.js'
export type { InsertQuery as Insert, InsertData } from './lib/InsertBuilder.js'
export { ASK } from './lib/AskBuilder.js'
export type { AskQuery as Ask } from './lib/AskBuilder.js'
export { DESCRIBE } from './lib/DescribeBuilder.js'
export type { DescribeQuery as Describe } from './lib/DescribeBuilder.js'
export { CONSTRUCT } from './lib/ConstructBuilder.js'
export type { ConstructQuery as Construct } from './lib/ConstructBuilder.js'
export { SELECT } from './lib/SelectBuilder.js'
export type { SelectQuery as Select } from './lib/SelectBuilder.js'
export { WITH } from './lib/WithBuilder.js'
export { prefixes } from '@tpluscode/rdf-string'
export type { SparqlTemplateResult } from '@tpluscode/rdf-string'
export type {
  SparqlAskExecutable, SparqlGraphQueryExecutable, SparqlQueryExecutable, SparqlUpdateExecutable,
} from './lib/index.js'
