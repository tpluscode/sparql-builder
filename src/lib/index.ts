import { AskQuery, ConstructQuery, QueryOptions, SelectQuery, UpdateQuery } from 'sparql-http-client'
import { SparqlTemplateResult } from '@tpluscode/rdf-string'
import prologue, { PrologueBuilder } from './partials/prologue'

interface SparqlBuildOptions {
  base?: string
}

export type SparqlExecuteOptions = QueryOptions & SparqlBuildOptions

export interface SparqlQuery extends PrologueBuilder {
  build(options?: SparqlBuildOptions): string
  _getTemplateResult(): SparqlTemplateResult
}

export interface SparqlQueryExecutable {
  execute<TQuery extends SelectQuery<any>>(client: TQuery, requestInit?: SparqlExecuteOptions): ReturnType<TQuery['select']>
}

export interface SparqlGraphQueryExecutable {
  execute<TQuery extends ConstructQuery<any>>(client: TQuery, requestInit?: SparqlExecuteOptions): ReturnType<TQuery['construct']>
}

export interface SparqlUpdateExecutable {
  execute<TQuery extends UpdateQuery<any>>(client: TQuery, requestInit?: SparqlExecuteOptions): ReturnType<TQuery['update']>
}

export interface SparqlAskExecutable {
  execute<TQuery extends AskQuery<any>>(client: TQuery, requestInit?: SparqlExecuteOptions): ReturnType<TQuery['ask']>
}

type TBuilder = Pick<SparqlQuery, 'build'> & Pick<SparqlTemplateResult, '_toPartialString'>

// eslint-disable-next-line no-unused-vars
export default function Builder<T extends SparqlQuery>(): TBuilder & T {
  return {
    ...prologue(),
    build(this: SparqlQuery, { base }: SparqlBuildOptions = {}): string {
      const queryResult = this._getTemplateResult().toString({
        base,
      })

      if (this.prologueResult) {
        return `${this.prologueResult}\n\n${queryResult}`
      }

      return queryResult
    },
    _toPartialString(this: SparqlQuery, options: any) {
      return this._getTemplateResult()._toPartialString(options)
    },
  } as any
}
