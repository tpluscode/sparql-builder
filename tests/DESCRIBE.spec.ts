import namespace from '@rdfjs/namespace'
import * as RDF from '@rdfjs/data-model'
import { namedNode } from '@rdfjs/data-model'
import { DESCRIBE } from '../src'
import { sparqlClient } from './_mocks'

describe('DESCRIBE', () => {
  it('executes a construct', async () => {
    // given
    const client = sparqlClient()

    // when
    await DESCRIBE``.execute(client)

    // then
    expect(client.construct).toHaveBeenCalled()
  })

  it('builds a DESCRIBE without WHERE', () => {
    // given
    const expected = 'DESCRIBE <urn:foo:bar>'

    // when
    const actual = DESCRIBE`${RDF.namedNode('urn:foo:bar')}`.build()

    // then
    expect(actual).toMatchQuery(expected)
  })

  it('can have additional prologue', () => {
    // given
    const base = namedNode('http://foo.bar/baz')

    // when
    const query = DESCRIBE`?foo`
      .prologue`#pragma join.hash off`
      .prologue`BASE ${base}`
      .WHERE`?foo a ?bar`
      .build()

    // then
    expect(query).toMatchSnapshot()
  })

  it('supports LIMIT/OFFSET', () => {
    // given
    const expected = 'DESCRIBE ?foo LIMIT 100 OFFSET 200'

    // when
    const actual = DESCRIBE`${RDF.variable('foo')}`.LIMIT(100).OFFSET(200).build()

    // then
    expect(actual).toMatchQuery(expected)
  })

  it('supports ORDER BY', () => {
    // given
    const expected = 'DESCRIBE ?foo WHERE { ?foo a ?bar } ORDER BY ?bar LIMIT 100 OFFSET 200'
    const foo = RDF.variable('foo')
    const bar = RDF.variable('bar')

    // when
    const actual = DESCRIBE`${foo}`
      .WHERE`${foo} a ${bar}`
      .ORDER().BY(bar)
      .LIMIT(100)
      .OFFSET(200)
      .build()

    // then
    expect(actual).toMatchQuery(expected)
  })

  it('supports FROM (NAMED)', () => {
    // given
    const expected = `DESCRIBE ?foo
      FROM <http://example.com/foo>
      FROM NAMED <http://example.com/bar>`

    // when
    const actual = DESCRIBE`${RDF.variable('foo')}`
      .FROM(RDF.namedNode('http://example.com/foo'))
      .FROM().NAMED(RDF.namedNode('http://example.com/bar')).build()

    // then
    expect(actual).toMatchQuery(expected)
  })

  it('can be constructed with a base', () => {
    // given
    const ns = namespace('http://example.com/')
    const expected = `BASE <http://example.com/>

    DESCRIBE <person>`

    // when
    const actual = DESCRIBE`${ns.person}`.build({
      base: 'http://example.com/',
    })

    // then
    expect(actual).toMatchQuery(expected)
  })
})
