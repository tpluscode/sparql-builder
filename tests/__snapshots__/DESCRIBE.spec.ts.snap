// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DESCRIBE can have additional prologue 1`] = `
"#pragma join.hash off
BASE <http://foo.bar/baz>

DESCRIBE ?foo

WHERE {
?foo a ?bar
}

"
`;
